### ISSUE TITLE
How can I provide an extended export ?

### ISSUE THREAD

**Issue by Riduidel** [shaarli/Shaarli#1450]
Current export format doesn't include the shaare permalink, neither the image.
I would like to have those infos in my export.

I was envisionning creating a plugin, but it seems like a plugin can't provide a new route, so I won't be able to have a new export route. So, what could I do to have such an extended export available without sacrifying the standard shaarli export ?

**Comment by nodiscc** [comment]
I don't think the "Netscape"-style HTML bookmarks format provides a way to show such information (actually I don't think there is a specification at all). Can you give an example snippet of what you would expect inside the HTML export?

**Comment by Riduidel** [comment]
I had discussions long ago with @Sebasauvage about the why of this export.

It is specified "somewhere" - but I can't find back where. Seems like
there are some infos scattered across the web :

http://fileformats.archiveteam.org/wiki/Netscape_bookmarks

http://forums.mozillazine.org/viewtopic.php?f=7&t=372993

Anyway, an example of what I want would be something like


       <dt>
         <a href="https://github.com/shaarli/Shaarli/issues/1450"
add_date="1588185934" private="0" tags="shaarli,plugin,system">
           How can I provide an extended export ? · Issue #1450 ·
shaarli/Shaarli
         </a>
       </dt>
       <dd>

          <a href="https://nicolas-delsaux.hd.free.fr/Shaarli/?MC1BQg"
class="permalink">

          <img class="thumbnail"
src="https://nicolas-delsaux.hd.free.fr/Shaarli/cache/thumb/99cd2175108d157588c04758296d1cfc/5f893917a7047f013f97e5fad3e228e42c422f53125901.jpg"/>

          </a>

          Bon ben finalement, ce plugin Shaarli, ça marche pas trop en
fait 😥

       </dd>

This would break the import, but as I use this export only to fill a
system where I can manage that extension, it would be no problem for me.
And that's why I thought initially about a plugin. But plugins can't add
new routes, and that's why I'm stuck.

Le 01/05/2020 à 16:39, nodiscc a écrit :

**Comment by Riduidel** [comment]
So here is what I did

I changed the `export.bookmarks.tpl` file to

```<!DOCTYPE NETSCAPE-Bookmark-file-1>
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
<!-- This is an automatically generated file.
     It will be read and overwritten.
     Do Not Edit! -->{ignore}The RainTPL loop is formatted to avoid generating extra newlines{/ignore}
<TITLE>{$pagetitle}</TITLE>
<H1>Shaarli export of {$selection} bookmarks on {$date}</H1>
<DL><p>{loop="links"}
<DT><A HREF="{$value.url}" ADD_DATE="{$value.timestamp}" PRIVATE="{$value.private}" TAGS="{$value.taglist}">{$value.title}</A>{if="$value.description"}{$eol}
<DD>
{$value.description}
<A HREF="?{$value.shorturl}">
<IMG SRC="{$value.thumbnail}"/>
</A>
{/if}{/loop}
</DL><p>
```

(notice I added the link and image after description, to minimize bad rendering if using import to import in a browser).

This way, my export contains everything I need ... but it's obviously only a hack.

**Comment by nodiscc** [comment]
So you add the permalink and thumbnail in the description field? Why not - it may be useful and doesn't break anything. I think we could provide an option/checkbox in the export dialog `Embed permalink and thumbnail in descriptions`?

**Comment by Riduidel** [comment]
Sorry for the late reply, I will work on it and provide a PR soon

**Comment by nodiscc** [comment]
Pull Request at https://github.com/shaarli/Shaarli/pull/1470

**Comment by ArthurHoaro** [comment]
It's my bad, I didn't review the latest changes and since it has no milestone, never came back to it.

**Comment by ArthurHoaro** [comment]
I ran a few tests and I don't think it's a good idea to extend the netscape format with some additional `<A>` field: it doesn't work well when it's imported back into Shaarli, it doesn't work when it's imported into a browser. So it kinf of defeat the purpose of exporting the bookmarks into this format.

I suggest that, instead, we close this, and we explore more in depth a way to export the entire DB into an easier format to use programmatically, such as #1823. Any thought?

**Comment by nodiscc** [comment]
I agree that HTML exports should stick to the basic netscape format as much as possible. Full JSON exports are already possible using the [API](https://shaarli.github.io/api-documentation/). Maybe we should work on a pure PHP export functionality which doesn't require third-party API clients? Using curl is another option, but not straightforward as it requires fetching and reusing a JWT token.

**Comment by virtadpt** [comment]
It would certainly be handy, @nodiscc.
